plugins/out_s3/s3.c

```diff
@@ -423,6 +485,67 @@ static int cb_s3_init(struct flb_output_instance *ins,
             flb_plg_error(ctx->ins, "'s3_key_format' must start with a '/'");
             return -1;
         }
+        if (strstr((char *) tmp, "$INDEX")) {
+            ctx->key_fmt_has_seq_index = FLB_TRUE;
+
+            /* Initialize local storage to store metadata files */
+            ret = s3_store_init(ctx);
+            if (ret == -1) {
+                flb_plg_error(ctx->ins, "Failed to initialize S3 storage: %s",
+                              ctx->store_dir);
+                return -1;
+            }
+
+            fs_stream = flb_fstore_stream_create(ctx->fs, "metadata");
+            if (!fs_stream) {
+                flb_plg_error(ctx->ins, "could not initialize metadata stream");
+                flb_fstore_destroy(ctx->fs);
+                ctx->fs = NULL;
+                return -1;
+            }
+
+            /* Construct directories and file path names */
+            tmp = fs_stream->path;
+            ctx->metadata_dir = flb_sds_create(tmp);
+            tmp = "/index_metadata";
+            ctx->metadata_dir = flb_sds_cat(ctx->metadata_dir, tmp, strlen(tmp));
+
+            ctx->seq_index_file = flb_sds_create(ctx->metadata_dir);
+            tmp = "/seq_index_";
+            ctx->seq_index_file = flb_sds_cat(ctx->seq_index_file, tmp, strlen(tmp));
+            tmp = (char *) &ctx->ins->name;
+            ctx->seq_index_file = flb_sds_cat(ctx->seq_index_file, tmp, strlen(tmp));
+
+            /* Create directory path if it doesn't exist */
+            ret = mkdir(ctx->metadata_dir, 0755);
+            if (ret < 0 && errno != EEXIST) {
+                flb_plg_error(ctx->ins, "Failed to create metadata directory");
+                return -1;
+            }
+
+            /* Check if index file doesn't exist and set index value */
+            if (access(ctx->seq_index_file, F_OK) != 0) {
+                ctx->seq_index = 0;
+                ctx->decrement_index = 0;
+                ret = write_seq_index(ctx->seq_index_file, ctx->seq_index,
+                                      ctx->decrement_index);
+                if (ret < 0) {
+                    flb_plg_error(ctx->ins, "Failed to write to metadata file");
+                    return -1;
+                }
+            } else {
+                ret = read_seq_index(ctx->seq_index_file, &ctx->seq_index, &ctx->decrement_index);
+                if (ret < 0) {
+                    flb_plg_error(ctx->ins, "Failed to read from metadata file");
```

Comment:
this error message and the one above could be better, lets explicitly say sequence file. When users complain of problems running fluent bit... its good to have very descriptive and unique error messages to help them & us get an idea of what went wrong